internal/kubernetes/controller/controller_builder.go

```diff
@@ -0,0 +1,87 @@
+package controller
+
+import (
+	"fmt"
+	pb "github.com/SAP/remote-work-processor/build/proto/generated"
+	"github.com/SAP/remote-work-processor/internal/grpc"
+	"github.com/SAP/remote-work-processor/internal/kubernetes/selector"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+type ControllerBuilder struct {
+	resource                      *pb.Resource
+	selector                      *selector.Selector
+	manager                       *Manager
+	reconciliationPeriodInMinutes int32
+}
+
+func NewControllerFor(r *pb.Resource) *ControllerBuilder {
+	return &ControllerBuilder{
+		resource: r,
+	}
+}
+
+func (c *ControllerBuilder) WithReconcilicationPeriodInMinutes(period int32) *ControllerBuilder {
+	c.reconciliationPeriodInMinutes = period
+	return c
+}
+
+func (c *ControllerBuilder) WithSelector(selector *selector.Selector) *ControllerBuilder {
+	c.selector = selector
+	return c
+}
+
+func (c *ControllerBuilder) ManagedBy(manager *Manager) *ControllerBuilder {
+	c.manager = manager
+	return c
+}
```

Comment:
Again, are these fields really optional or are they always required in order for the code to work?

If they are always required, I dislike the idea of having them as optional parameters just so we save some horizontal space (in return for using more vertical one) and introducing potential future runtime errors.

Reply:
What would you say to adding a validation in the `build()` method of all builders that checks for the required fields, and if any aren't set to return an error? This would be semantically similar to the [Immutables](https://immutables.github.io/) library for Java.

Reply:
It would be an improvement, but it would still be a runtime check rather than a compile time error. My main gripe with required "optionals" is that you have no compile time information that you forgot to set a field or not. If you're worried about the "constructor" function having too much horizontal space taken, you can always format it as:

```
func example(
	firstRequired string,
	secondRequired int,
	thirdRequired byte,
	optionals ...Option
) {
<code>
}
```